### PR TITLE
MCOL-4188 Regression fixes for MCOL-641.

### PR DESCRIPTION
**Pull request by tntnatbry**
1. In BatchPrimitiveProcessorJL::countThisMsg(), exit early if the
return code from primproc is non-zero.
2. For statistical window functions (stddev/var), properly handle
the case when there is only one value for the calculation.

### CODE REVIEW

File: dbcon/joblist/batchprimitiveprocessor-jl.cpp [hunk 1]
```diff
@@ -703,6 +703,12 @@ bool BatchPrimitiveProcessorJL::countThisMsg(messageqcpp::ByteStream& in) const
     const uint8_t* data = in.buf();
     uint32_t offset = sizeof(ISMPacketHeader) + sizeof(PrimitiveHeader); // skip the headers
 
+    ISMPacketHeader* hdr = (ISMPacketHeader*)(data);
+
+    // Exit early if PrimProc has thrown an exception
+    if (hdr->Status > 0)
+        return true;
```
Comment (drrtuy) [inline dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: This is a very good change but you should also log a message explaining that PP throws an exception running a primitive.

File: utils/windowfunction/wf_stats.cpp [hunk 2]
```diff
@@ -225,9 +224,17 @@ void WF_stats<T>::operator()(int64_t b, int64_t e, int64_t c)
     {
         setValue(CalpontSystemCatalog::DOUBLE, b, e, c, (double*) NULL);
     }
-    else if (fCount == 1 && (fFunctionId == WF__STDDEV_SAMP || fFunctionId == WF__VAR_SAMP))
+    else if (fCount == 1)
     {
-        setValue(CalpontSystemCatalog::DOUBLE, b, e, c, (double*) NULL);
+        if (fFunctionId == WF__STDDEV_SAMP || fFunctionId == WF__VAR_SAMP)
```
Comment (drrtuy) [inline utils/windowfunction/wf_stats.cpp:229]: I believe we are fixing something that is outside of the scope of 4188 here or am I wrong?

File: dbcon/joblist/batchprimitiveprocessor-jl.cpp (hunk 1 above)
Comment (tntnatbry) [reply dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: This check is added because the below assertion in this function `idbassert(in.length() > offset);` was failing when the buffer contained an error message. We need to skip the rest of this function execution in case of an error.

The error logging is handled by the calling function TupleBPS::receiveMultiPrimitiveMessages. Specifically this code block:

```
                if (bs->length() == 0 || hdr->Status > 0)
                {
                    /* PM errors mean this should abort right away instead of draining the PM backlog */
                    tplLock.lock();

                    if (bs->length() == 0)
                    {
                        errorMessage(IDBErrorInfo::instance()->errorMsg(ERR_PRIMPROC_DOWN));
                        status(ERR_PRIMPROC_DOWN);
                    }
                    else
                    {
                        string errMsg;

                        bs->advance(sizeof(ISMPacketHeader) + sizeof(PrimitiveHeader));
                        *bs >> errMsg;
                        status(hdr->Status);
                        errorMessage(errMsg);
                    }

                    abort_nolock();
                    goto out;
                }
```

File: utils/windowfunction/wf_stats.cpp (hunk 2 above)
Comment (tntnatbry) [reply utils/windowfunction/wf_stats.cpp:229]: This change is fixing a regression in working_tpch1_compareLogOnly/windowFunctions/MCOL-3423.sql. I did not see this test failing in develop prior to the MCOL-641 merge. So it's possible some change done to this file as part of MCOL-641 introduced this regression. Now, we are handling `fCount == 0` and `fCount == 1` here as it is handled in the non-window function versions of these functions in `RowAggregationUM::calculateStatisticsFunctions`

File: dbcon/joblist/batchprimitiveprocessor-jl.cpp (hunk 1 above)
Comment (drrtuy) [reply dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: If PP has thrown an exception there will be no error message from PP and it looks like an assertion in ExeMgr w/o any note about PP. That is why I asked about an explicit message.

Comment (tntnatbry) [reply dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: This is not true. PP sends the error info (code and message) to ExeMgr in case an exception is raised by PP. Refer to the catch blocks in `BatchPrimitiveProcessor::execute()`. These blocks call `writeErrorMsg` function which log the error info to bytestream, which are then read by the ExeMgr on the receiving side.

Comment (drrtuy) [reply dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: I suggest you to try to throw an exception from BatchPrimitiveProcessor::execute() and see how the logging output helps you to understand that PP throws.

Comment (drrtuy) [reply dbcon/joblist/batchprimitiveprocessor-jl.cpp:710]: I believe that was a good moment to finally start logging something usefull in case of PP throw. I believe I should do it on my own.

**Review by drrtuy — APPROVED** [review]
